### ISSUE TITLE
vararg ReactElement implicitly converted?

### ISSUE THREAD

**Issue by evbo** [shadaj/slinky#282]
Slinky version `0.6.2`

possibly a bug that was reported closed:
https://github.com/shadaj/slinky/issues/245
https://github.com/shadaj/slinky/issues/185
https://github.com/shadaj/slinky/issues/239

Let's say I have a function that returns a `ReactElement` and takes a `ReactElement*`:

```scala
def dataMessage(message: ReactElement*): = {
    div(className := "someData")(message)
  }
```

Ideally I'd send it a list of content, like:

```scala
dataMessage(List("testElement ", i(className := "icon"), " other text"): _*)
```

But this won't compile unless I force it with:

```scala
dataMessage(List("testElement ": ReactElement, i(className := "icon"): ReactElement, " other text": ReactElement): _*)
```

And even still, I then get warning: 

> Warning: Each child in a list should have a unique "key" prop.

My workaround is to call it like:

```scala
dataMessage(div(key := "message")("testElement ", i(className := "icon"), " other text"))
```

which adds an unnecessary div but at least allows me to add the `key`

**Comment by baank** [comment]
So there already exists an implicit conversion between a Tag and a ReactElement.

But since you are using a List it doesn't apply. One option is to have something like this:
```
implicit def convert(l: List[WithAttrs[_]])(implicit ev: WithAttrs[_] => ReactElement): List[ReactElement] = l map { a => a: ReactElement }
```

@shadaj .. What do you think about having implicit conversions for Option, Seq, Map etc built into the TagComponent class. I'm not sure what other projects do to handle this situation.

**Comment by shadaj** [comment]
So for the first compilation issue, the recommended solution would be to construct the list as `List[ReactElement](...)` so that the individual implicit conversions aren't required. As I mentioned in #283, I'm not super comfortable with implicit conversions that would fix the compile error you saw because of the potential for performance degradation with unexpected magic.

For the key error, that looks like an issue in `dataMessage`, where you should be passing in `message` as `message: _*` since you want to expand out those children instead of passing them to React as a list.

**Comment by baank** [comment]
@shadaj .. I added the following to TagComponent and it fixes this and my problem without breaking any of the tests. There should be no different in performance since when you specify the type the implicit conversion will still happen for each of the elements of the List anyway. But I appreciate your concerns over the unexpected magic. I would be happy to submit a PR and add additional tests if you reconsider.

```
  @inline implicit def build(withAttrs: Seq[WithAttrs[_]]): Seq[ReactElement] =
    withAttrs.map(build)
```

**Comment by evbo** [comment]
@shadaj in my example I was passing them in as: `: _*`

The only time I wasn't was for the "workaround", but that's because I'm using a div, not list in that case

**Comment by shadaj** [comment]
@evbo so, you still get the key error with this?
```scala
def dataMessage(message: ReactElement*): = {
  div(className := "someData")(message: _*)
}
```

@baank hmm, I'm starting to lean towards including these implicits, since the experience without them is pretty confusing. If we do include this, we should try to see if we can make the `Seq` conversion more generic, so that we handle `List => List`, `Array => Array`, etc.

**Comment by evbo** [comment]
@shadaj interesting, For the simple case of a single `div`, I appear to need deconstruction both with the `ReactElement*` argument and when applying argument to the `div` inside the `dataMessage` method:

```scala
dataMessage(
        Seq[ReactElement](
          div("hey")                  
        ): _*                                   //<------ : _* necessary here too
)

def dataMessage(message: ReactElement*): = {
  div(className := "someData")(message: _*)     //<------ : _* necessary here like you pointed out
}
```

However, there's one other caveat I've noticed: You need to wrap any additional logic in a function. For instance, if you try to add an if statement it gives: 
> no `: _*' annotation allowed here

```scala
val branch = true

dataMessage(
  if (branch){
    Seq[ReactElement](
      div("hey")
    ): _*
  } else {
    Seq[ReactElement](
      div("bye")
    ): _*
  }
)

```

So instead it takes just a little extra code to work:

```scala
dataMessage(
  getBranch(branch): _*
)

def getBranch(branch: Boolean): Seq[ReactElement] = {
  if (branch){
    Seq[ReactElement](
      div("hey")
    )
  } else {
    Seq[ReactElement](
      div("bye")
    )
  }
}
```

**Comment by baank** [comment]
@shadaj  .. We should be able to do this as a higher kinded type. Will submit a PR and also add a test for @evbo's issue.

**Comment by baank** [comment]
@evbo @shadaj .. Pull request submitted which should fix this (and many other similar issues) #285

**Comment by evbo** [comment]
@shadaj working great for my use-cases above on `0.6.2+6-cc9d534e`, feel free to close :+1:

**Comment by shadaj** [comment]
Fixed by #285!
